### ISSUE TITLE
XaAES doesn't offer resolution change application

### ISSUE THREAD

**Issue by mikrosk** [freemint/freemint#381]
Another input from @swe-yc: there isn't a resolution change dialog, like N.AES has. Interestingly, there definitely used to be at least some kind of stub, see: http://xaaes.gokmase.com/docs.php?chapter=6.htm#6.6.1 but for some reason it was never officially enabled and then removed.

Teradesk's resolution change dialog helps a bit but still, e.g. Thing needs this as an external application.

**Comment by skaftetryne** [comment]
There isn't a resolution changer in N.AES either, it is taken care of by a separate application.

**Comment by th-otto** [comment]
It's not the task of AES to supply such an dialog. It always depends on the VDI in use, and all other desktops also use an external application for this purpose.

**Comment by GokMasE** [comment]
> Another input from @swe-yc: there isn't a resolution change dialog, like N.AES has. Interestingly, there definitely used to be at least some kind of stub, see: http://xaaes.gokmase.com/docs.php?chapter=6.htm#6.6.1 but for some reason it was never officially enabled and then removed.

Yes, I can say for sure there has been such a dialog present - and that it worked! I know, since I took that screenshot back in the days :)
 It was indeed also possible to set current resolution to trigger a plain restart of the AES.

If there are ideological objections to making such a thing re-appear within XaAES (although it was definitely very handy), I'd say it should (for someone that knows his way around Pure-C/GCC)  not be a huge task to look into turning the old code (which I presume is still archived somewhere?) into a tiny standalone binary. This could then be called from the system menu bar of XaAES (or by a keyboard shortcut) so that to the user it would not be any different than using such a function built-in.

**Comment by mikrosk** [comment]
> Ok but does the one from N.AES work in XaAES? That would be quite surprising. 

Clarification from Swe: _Yes, and it works. There is some complainig from XaAES, but it works._ Maybe those warnings are similar as @th-otto fixed in https://github.com/freemint/freemint/commit/8ee8594b8ea00615ce98706aaa480f5880671849.

**Comment by skaftetryne** [comment]
> If there are ideological objections to making such a thing re-appear within XaAES (although it was definitely very handy), I'd say it should (for someone that knows his way around Pure-C/GCC) not be a huge task to look into turning the old code (which I presume is still archived somewhere?) into a tiny standalone binary. This could then be called from the system menu bar of XaAES (or by a keyboard shortcut) so that to the user it would not be any different than using such a function built-in.

Not so much ideology as practicality. You need specific support for lots of different hardware, might even be easier to write specific applications for the various graphics cards.

The video mode selector for XaAES that I made for the Firebee and adapted for VanillaMiNT works with ST, TT and Falcon stock video. It also works with FireTOS XBIOS (on the Firebee) and the SuperVidel. It's already getting a bit complicated, with different dialogs for each hardware. E.g. FireTOS allows the client to retrieve a list of available resolutions through XBIOS, SuperVidel does not and need a hardcoded list. FireTOS XBIOS extensions are "inspired" by Milan TOS but it not compatible, and it's not compatible with itself on different hardware (Firebee Videl vs Radeon). Falcon videomodes are flexible, ST and TT not. And so on...  I don't think it makes much sense to do this *inside* XaAES.

N.AES use the environment variable SDMASTER which points to the resolution changer binary, which is then ran as soon as the user wants to change resolution. It should be trivial to add the same to XaAES.

**Comment by th-otto** [comment]
> You need specific support for lots of different hardware, might even be easier to write specific applications for the various graphics cards.

IIRC, there are even such applications for specific hardware like NOVA graphic cards. I totally agree that such functionality does not belong into AES. Maybe (as you already pointed out) it is even impossible  to write a single application that supports all possible hardware.

SDMASTER is therefor also a thing that does not belong into the AES configuration, it is the desktop that should check for that variable (if it offers a menu item to change the resolution, teradesk only handles the standard ST resolutions).

**Comment by skaftetryne** [comment]
> SDMASTER is therefor also a thing that does not belong into the AES configuration

If you want to initiate a resolution change from the XaAES system menu (as GokMasE suggested) RSMASTER would be an easy way to specify which binary to use.

> it is the desktop that should check for that variable (if it offers a menu item to change the resolution, teradesk only handles the standard ST resolutions).

I agree that the desktop is the most obvious place for this. Teradesk can already call an external resolution changer. Can't remember if it evaluates RSMASTER, but you can install an application specifically for changing resolution.

**Comment by mikrosk** [comment]
>  teradesk only handles the standard ST resolutions

Surely not, it offers both TT and Falcon resolutions.

As for the rest, I'm totally fine if we have _some_ solution at least for the base ST/TT/Falcon models, even an external one. That was Swe's point: he's actively using Falcon so he appreciate the option to switch to mono / 16 / 256 colours when desired.

**Comment by GokMasE** [comment]
Yes, I think it would be helpful with some kind of solution too - I would be fine with an external solution.
If a standalone solution can be called with little extra overhead I can't see why not. 

While I understand the logic some ppl see in having this handled by desktop, I don't feel that way myself.

Desktops like Thing offer so very much functionality that customize the way the system interacts with the user that I more and more feel the resolution dialog is more logical to address from the xaaes menu bar than the desktops. To me the desktop is typically managing higher level issues and ordering a resolution switch feels more logical to me if done from the core/aes of the system.

**Comment by th-otto** [comment]
For me, it is also a matter of consistency. If the desktop already has a dialog to change the resolution (and i think most of them do, either builtin, or by calling an external application), it does not make sense (and is also confusing) to implement that also in some AES menu.

**Comment by mikrosk** [comment]
Just for the record, the resolution change dialog disappeared by introducing 1.17 branch:
- EasyMiNT's XaAES from May 6 2007 and the dialog is there and working nicely
- the last 1.16 xaaes commit is bfe8d362 (can't compile without gcc2) dated 13.05.2007
- the first 1.17 xaaes commit is 62947f1e and clicking on the Process menu entry or the button in Task Manager doesn't do anything
- 1.17.0 release has the Process menu entry greyed out and the button in Task Manager is gone

So yeah, it would seem that Ozk came to the same conclusion as a farewell gift before his departure he deactivated the resolution change dialog. ;)
